File: src/ai/providers/anthropic.ts
/**
 * Anthropic provider adapter.
 * Supports Claude Opus, Sonnet, and Haiku model families.
 *
 * Endpoint: https://api.anthropic.com/v1/messages
 * Auth:     x-api-key header + anthropic-version: 2023-06-01
 * System:   Top-level "system" field, NOT in messages array
 * Thinking: thinking.type: "enabled", thinking.budget_tokens
 * Stream:   event: content_block_delta
 *           data: {"type":"content_block_delta","delta":{"type":"text_delta","text":"..."}}
 */

import type { ChatMessage, ChatOptions, StreamEvent } from "../types";
import { BaseProvider } from "./base";

const ENDPOINT = "https://api.anthropic.com/v1/messages";
const API_VERSION = "2023-06-01";

export class AnthropicProvider extends BaseProvider {
  async chat(
    model: string,
    messages: ChatMessage[],
    options?: ChatOptions
  ): Promise<ChatMessage> {
    const body = this.buildRequestBody(model, messages, options, false);

    const response = await fetch(ENDPOINT, {
      method: "POST",
      headers: this.buildHeaders(),
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(
        `Anthropic request failed (${response.status}): ${errorText}`
      );
    }

    const data = await response.json();

    let content = "";
    let thinking = "";

    if (Array.isArray(data.content)) {
      for (const block of data.content) {
        if (block.type === "text") {
          content += block.text;
        } else if (block.type === "thinking") {
          thinking += block.thinking;
        }
      }
    }

    return {
      role: "assistant",
      content,
      thinking: thinking || undefined,
      timestamp: Date.now(),
    };
  }

  async *chatStream(
    model: string,
    messages: ChatMessage[],
    options?: ChatOptions
  ): AsyncGenerator<StreamEvent> {
    const body = this.buildRequestBody(model, messages, options, true);

    const response = await fetch(ENDPOINT, {
      method: "POST",
      headers: this.buildHeaders(),
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      yield {
        type: "error",
        message: `Anthropic stream failed (${response.status}): ${errorText}`,
      };
      return;
    }

    const reader = response.body?.getReader();
    if (!reader) {
      yield { type: "error", message: "No response body from Anthropic" };
      return;
    }

    let inputTokens = 0;
    let outputTokens = 0;

    // Anthropic uses a custom SSE format with "event:" lines followed by "data:" lines.
    // We parse both event type and data payload.
    const decoder = new TextDecoder();
    let buffer = "";
    let currentEvent = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";

      for (const line of lines) {
        const trimmed = line.trim();

        if (trimmed.startsWith("event: ")) {
          currentEvent = trimmed.slice(7);
          continue;
        }

        if (trimmed.startsWith("data: ")) {
          let data: Record<string, unknown>;
          try {
            data = JSON.parse(trimmed.slice(6));
          } catch {
            continue;
          }

          const streamEvent = this.handleSSEEvent(
            currentEvent,
            data
          );

          if (streamEvent) {
            yield streamEvent;
          }

          // Collect usage from message_delta event
          if (currentEvent === "message_delta" && data.usage) {
            const usage = data.usage as Record<string, number>;
            outputTokens = usage.output_tokens ?? outputTokens;
          }

          // Collect usage from message_start event
          if (currentEvent === "message_start" && data.message) {
            const msg = data.message as Record<string, unknown>;
            if (msg.usage) {
              const usage = msg.usage as Record<string, number>;
              inputTokens = usage.input_tokens ?? 0;
            }
          }

          currentEvent = "";
        }
      }
    }

    yield {
      type: "done",
      usage: { input_tokens: inputTokens, output_tokens: outputTokens },
    };
  }

  private handleSSEEvent(
    eventType: string,
    data: Record<string, unknown>
  ): StreamEvent | null {
    if (eventType === "content_block_delta") {
      const delta = data.delta as Record<string, string> | undefined;
      if (!delta) return null;

      if (delta.type === "thinking_delta" && delta.thinking) {
        return { type: "thinking", content: delta.thinking };
      }

      if (delta.type === "text_delta" && delta.text) {
        return { type: "content", content: delta.text };
      }
    }

    if (eventType === "error") {
      const error = data.error as Record<string, string> | undefined;
      return {
        type: "error",
        message: error?.message ?? "Unknown Anthropic stream error",
      };
    }

    return null;
  }

  private buildHeaders(): Record<string, string> {
    return {
      "Content-Type": "application/json",
      "x-api-key": this.apiKey,
      "anthropic-version": API_VERSION,
    };
  }

  private buildRequestBody(
    model: string,
    messages: ChatMessage[],
    options: ChatOptions | undefined,
    stream: boolean
  ): Record<string, unknown> {
    // Extract system prompt from messages — Anthropic expects it as a top-level field
    const systemMessages = messages.filter((m) => m.role === "system");
    const nonSystemMessages = messages.filter((m) => m.role !== "system");

    const body: Record<string, unknown> = {
      model,
      messages: this.formatMessages(nonSystemMessages),
      max_tokens: options?.maxTokens ?? 4096,
      stream,
    };

    // Set system prompt as top-level field
    if (systemMessages.length > 0) {
      body.system = systemMessages.map((m) => m.content).join("\n\n");
    }

    // Temperature (only for non-thinking requests)
    if (!options?.thinking && options?.temperature !== undefined) {
      body.temperature = options.temperature;
    }

    // Extended thinking configuration
    if (options?.thinking) {
      body.thinking = {
        type: "enabled",
        budget_tokens: Math.min(
          options.maxTokens ? Math.floor(options.maxTokens * 0.8) : 10000,
          100000
        ),
      };
    }

    return body;
  }

  private formatMessages(
    messages: ChatMessage[]
  ): Array<Record<string, unknown>> {
    return messages.map((m) => {
      // If the message has images, use multipart content format
      if (m.images && m.images.length > 0 && m.role === "user") {
        const content: Array<Record<string, unknown>> = [];

        for (const img of m.images) {
          // Determine media type from data URI or default to jpeg
          let mediaType = "image/jpeg";
          let base64Data = img;

          const dataUriMatch = img.match(
            /^data:(image\/[a-z+]+);base64,(.+)$/
          );
          if (dataUriMatch) {
            mediaType = dataUriMatch[1];
            base64Data = dataUriMatch[2];
          }

          content.push({
            type: "image",
            source: {
              type: "base64",
              media_type: mediaType,
              data: base64Data,
            },
          });
        }

        content.push({ type: "text", text: m.content });
        return { role: m.role, content };
      }

      return { role: m.role, content: m.content };
    });
  }
}

File: crates/provider-adapters/src/anthropic.rs
use serde_json::{json, Value};

use super::types::{ChatRequest, ChatResponse, StreamEvent, Usage};

const ANTHROPIC_API_URL: &str = "https://api.anthropic.com/v1/messages";
const ANTHROPIC_VERSION: &str = "2023-06-01";

/// Build the Anthropic Messages API request body from our unified ChatRequest.
/// Anthropic requires the system message in a separate top-level field and uses
/// `x-api-key` / `anthropic-version` headers instead of Bearer auth.
fn build_request_body(request: &ChatRequest) -> Value {
    let mut system_text: Option<String> = None;
    let mut messages: Vec<Value> = Vec::new();

    for msg in &request.messages {
        if msg.role == "system" {
            // Anthropic expects system as a separate top-level field.
            system_text = Some(msg.content.clone());
        } else {
            messages.push(json!({
                "role": msg.role,
                "content": msg.content,
            }));
        }
    }

    let mut body = json!({
        "model": request.model,
        "messages": messages,
        "max_tokens": request.max_tokens,
        "temperature": request.temperature,
        "stream": request.stream,
    });

    if let Some(sys) = system_text {
        body["system"] = json!(sys);
    }

    // Enable extended thinking if requested
    if request.thinking {
        body["thinking"] = json!({
            "type": "enabled",
            "budget_tokens": request.max_tokens.saturating_sub(100).max(1024)
        });
        // Anthropic does not allow temperature with thinking
        body.as_object_mut().unwrap().remove("temperature");
    }

    body
}

/// Send a non-streaming request to the Anthropic Messages API.
pub async fn send_request(
    client: &reqwest::Client,
    api_key: &str,
    request: &ChatRequest,
) -> Result<ChatResponse, String> {
    let mut body = build_request_body(request);
    body["stream"] = json!(false);

    let response = client
        .post(ANTHROPIC_API_URL)
        .header("x-api-key", api_key)
        .header("anthropic-version", ANTHROPIC_VERSION)
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Anthropic request failed: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response
            .text()
            .await
            .unwrap_or_else(|_| "unknown error".to_string());
        return Err(format!("Anthropic API error ({}): {}", status, text));
    }

    let data: Value = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse Anthropic response: {}", e))?;

    // Anthropic returns content as an array of content blocks.
    let mut content = String::new();
    let mut thinking = None;

    if let Some(blocks) = data["content"].as_array() {
        for block in blocks {
            match block["type"].as_str() {
                Some("text") => {
                    if let Some(text) = block["text"].as_str() {
                        content.push_str(text);
                    }
                }
                Some("thinking") => {
                    if let Some(text) = block["thinking"].as_str() {
                        thinking = Some(text.to_string());
                    }
                }
                _ => {}
            }
        }
    }

    let input_tokens = data["usage"]["input_tokens"].as_u64().unwrap_or(0) as u32;
    let output_tokens = data["usage"]["output_tokens"].as_u64().unwrap_or(0) as u32;

    Ok(ChatResponse {
        content,
        thinking,
        usage: Usage {
            input_tokens,
            output_tokens,
        },
    })
}

/// Send a streaming request to the Anthropic Messages API.
pub async fn send_stream(
    client: &reqwest::Client,
    api_key: &str,
    request: &ChatRequest,
) -> Result<reqwest::Response, String> {
    let mut body = build_request_body(request);
    body["stream"] = json!(true);

    let response = client
        .post(ANTHROPIC_API_URL)
        .header("x-api-key", api_key)
        .header("anthropic-version", ANTHROPIC_VERSION)
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Anthropic stream request failed: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response
            .text()
            .await
            .unwrap_or_else(|_| "unknown error".to_string());
        return Err(format!(
            "Anthropic stream API error ({}): {}",
            status, text
        ));
    }

    Ok(response)
}

/// Parse an SSE chunk from Anthropic's streaming format into unified StreamEvents.
///
/// Anthropic uses `event: <type>` lines followed by `data: <json>` lines.
/// Key event types: content_block_delta, message_delta, message_stop.
pub fn parse_stream_chunk(chunk: &str) -> Vec<StreamEvent> {
    let mut events = Vec::new();
    let mut current_event_type: Option<String> = None;

    for line in chunk.lines() {
        let line = line.trim();

        if line.starts_with("event: ") {
            current_event_type = Some(line[7..].to_string());
            continue;
        }

        if !line.starts_with("data: ") {
            continue;
        }

        let data = &line[6..];
        let event_name = current_event_type.take();

        if let Ok(parsed) = serde_json::from_str::<Value>(data) {
            match event_name.as_deref() {
                Some("content_block_delta") => {
                    if let Some(text) = parsed["delta"]["text"].as_str() {
                        if !text.is_empty() {
                            events.push(StreamEvent {
                                event_type: "content".to_string(),
                                content: Some(text.to_string()),
                                usage: None,
                                message: None,
                            });
                        }
                    }
                }
                Some("message_delta") => {
                    if let Some(usage_obj) = parsed.get("usage") {
                        if !usage_obj.is_null() {
                            let output_tokens =
                                usage_obj["output_tokens"].as_u64().unwrap_or(0) as u32;
                            events.push(StreamEvent {
                                event_type: "usage".to_string(),
                                content: None,
                                usage: Some(Usage {
                                    input_tokens: 0,
                                    output_tokens,
                                }),
                                message: None,
                            });
                        }
                    }
                }
                Some("message_start") => {
                    if let Some(usage_obj) = parsed["message"].get("usage") {
                        if !usage_obj.is_null() {
                            let input_tokens =
                                usage_obj["input_tokens"].as_u64().unwrap_or(0) as u32;
                            events.push(StreamEvent {
                                event_type: "usage".to_string(),
                                content: None,
                                usage: Some(Usage {
                                    input_tokens,
                                    output_tokens: 0,
                                }),
                                message: None,
                            });
                        }
                    }
                }
                Some("message_stop") => {
                    events.push(StreamEvent {
                        event_type: "done".to_string(),
                        content: None,
                        usage: None,
                        message: None,
                    });
                }
                _ => {}
            }
        }
    }

    events
}
